### PR TITLE
kvengine: refine table format

### PR DESCRIPTION
**Pull request by coocood**
### What is changed and how it works?

What's Changed:

```commit-message
Make the table format more extensible for future change.
```

### CODE REVIEW

File: components/kvengine/src/table/sstable/iterator.rs [hunk 1]
```diff
@@ -64,12 +65,13 @@ impl BlockIterator {
 
     fn load_entries(&mut self) {
         let data = self.b.chunk();
-        let num_entries = LittleEndian::read_u32(data) as usize;
+        assert_eq!(data[0], BLOCK_FORMAT_V1);
+        let num_entries = LittleEndian::read_u32(&data[1..]) as usize;
         self.entry_offs = unsafe {
-            let ptr = data[4..].as_ptr() as *mut u32;
+            let ptr = data[5..].as_ptr() as *mut u32;
```
Comment (youjiali1995) [inline components/kvengine/src/table/sstable/iterator.rs:71]: It's not memory-aligned. I'm afriad it has poor performance. I prefer to use `u32` as version.

BTW, `finish_block()` uses `put_u32_le()` to write offsets, why here casts it to u32 directly?

File: components/kvengine/src/table/sstable/builder.rs [hunk 2]
```diff
@@ -16,11 +16,13 @@ pub const PROP_KEY_MAX_TS: &str = "max_ts";
...
 pub const PROP_KEY_TOMBS: &str = "tombs";
-pub const PROP_KEY_FUSE8: &str = "fuse8";
+pub const AUX_INDEX_BINARY_FUSE8: u8 = 1;
+pub const INDEX_FORMAT_V1: u8 = 1;
+pub const BLOCK_FORMAT_V1: u8 = 1;
 pub const NO_COMPRESSION: u8 = 0;
 pub const LZ4_COMPRESSION: u8 = 1;
 pub const ZSTD_COMPRESSION: u8 = 2;
-const TABLE_FORMAT: u16 = 1;
+pub const TABLE_FORMAT_V1: u16 = 1;
```
Comment (youjiali1995) [inline components/kvengine/src/table/sstable/builder.rs:25]: Why not use table format only? as long as there is a format change, we increases it only.

Comment (coocood) [reply components/kvengine/src/table/sstable/builder.rs:25]: table format is on higher level, it's hard to describe the combination of each component's format.
So the table format will be used to define the layout of each components, each component can have there own format.

**Review by youjiali1995 — APPROVED** [review]

**Comment by youjiali1995** [comment]
Please `make format` @coocood 
```
[2022-05-11T07:14:43.681Z] --- a/components/kvengine/src/table/sstable/builder.rs

[2022-05-11T07:14:43.681Z] +++ b/components/kvengine/src/table/sstable/builder.rs

[2022-05-11T07:14:43.681Z] @@ -235,7 +235,7 @@ impl Builder {

[2022-05-11T07:14:43.681Z]          }

[2022-05-11T07:14:43.681Z]      }

[2022-05-11T07:14:43.681Z]  

[2022-05-11T07:14:43.681Z] -    fn build_aux_index(&self, buf: &mut BytesMut, fuse8: &[u8] ) {

[2022-05-11T07:14:43.681Z] +    fn build_aux_index(&self, buf: &mut BytesMut, fuse8: &[u8]) {

[2022-05-11T07:14:43.681Z]          let origin_len = buf.len();

[2022-05-11T07:14:43.681Z]          buf.put_u32_le(0);

[2022-05-11T07:14:43.681Z]          buf.put_u32_le(AUX_INDEX_BINARY_FUSE8);

[2022-05-11T07:14:43.681Z] diff --git a/components/kvengine/src/table/sstable/iterator.rs b/components/kvengine/src/table/sstable/iterator.rs

[2022-05-11T07:14:43.681Z] index 3b8ab39b3..c3e6598d9 100644

[2022-05-11T07:14:43.681Z] --- a/components/kvengine/src/table/sstable/iterator.rs

[2022-05-11T07:14:43.681Z] +++ b/components/kvengine/src/table/sstable/iterator.rs

[2022-05-11T07:14:43.681Z] @@ -6,8 +6,7 @@ use byteorder::{ByteOrder, LittleEndian};

[2022-05-11T07:14:43.681Z]  use bytes::{Buf, Bytes, BytesMut};

[2022-05-11T07:14:43.681Z]  

[2022-05-11T07:14:43.681Z]  use super::{builder::META_HAS_OLD, SSTable};

[2022-05-11T07:14:43.681Z] -use crate::table::{search, table, LocalAddr};

[2022-05-11T07:14:43.681Z] -use crate::table::sstable::BLOCK_FORMAT_V1;

[2022-05-11T07:14:43.681Z] +use crate::table::{search, sstable::BLOCK_FORMAT_V1, table, LocalAddr};

[2022-05-11T07:14:43.681Z]  

[2022-05-11T07:14:43.681Z]  pub struct BlockIterator {

[2022-05-11T07:14:43.681Z]      b: Bytes,

[2022-05-11T07:14:43.681Z] diff --git a/components/kvengine/src/table/sstable/sstable.rs b/components/kvengine/src/table/sstable/sstable.rs

[2022-05-11T07:14:43.681Z] index c5e902148..cec35fec7 100644

[2022-05-11T07:14:43.681Z] --- a/components/kvengine/src/table/sstable/sstable.rs

[2022-05-11T07:14:43.681Z] +++ b/components/kvengine/src/table/sstable/sstable.rs

[2022-05-11T07:14:43.681Z] @@ -163,7 +163,10 @@ impl SSTableCore {

[2022-05-11T07:14:43.681Z]              let mut aux_idx_slice = aux_idx_data.chunk();

[2022-05-11T07:14:43.681Z]              validate_checksum(aux_idx_slice, footer.checksum_type)?;

[2022-05-11T07:14:43.681Z]              aux_idx_slice = &aux_idx_slice[4..];

[2022-05-11T07:14:43.681Z] -            assert_eq!(LittleEndian::read_u32(aux_idx_slice), AUX_INDEX_BINARY_FUSE8);

[2022-05-11T07:14:43.681Z] +            assert_eq!(

[2022-05-11T07:14:43.681Z] +                LittleEndian::read_u32(aux_idx_slice),

[2022-05-11T07:14:43.681Z] +                AUX_INDEX_BINARY_FUSE8

[2022-05-11T07:14:43.681Z] +            );

[2022-05-11T07:14:43.681Z]              aux_idx_slice = &aux_idx_slice[4..];

[2022-05-11T07:14:43.681Z]              let fuse8_len = LittleEndian::read_u32(aux_idx_slice) as usize;

[2022-05-11T07:14:43.681Z]              aux_idx_slice = &aux_idx_slice[4..];

[2022-05-11T07:14:43.681Z] + echo Please make format and run tests before creating a PR

[2022-05-11T07:14:43.681Z] Please make format and run tests before creating a PR

[2022-05-11T07:14:43.681Z] + exit 1
```

**Pull request merged**